### PR TITLE
Add support for 0d NaN / NaT Times

### PR DESCRIPTION
**Pull request by maxnoe**
# Description

For further discussion relevant to #6509, I here add test cases for possible `nan` / `NaT` values for `astropy.time` objects.

I have also started to look into implementing this.

### CODE REVIEW

File: astropy/time/tests/test_nan.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    'iso': '2020-01-01 20:00:00.5',
+    'isot': '2020-01-01T20:00:00.5',
+    'yday': '2020:300:20:00:05',
+}
+
+
+def test_time_creation():
+    '''Test a time scalar can be created as nan'''
+    t = Time(np.nan, format='jd')  # for now, in principle, it shouldn't matter
+    assert np.isnan(t._time.jd2)
```
Comment (taldcroft) [inline astropy/time/tests/test_nan.py:33]: Storing NaN in jd2 is considered an internal implementation detail of masking and could (in theory) change.  So I'd recommend not testing this here.

**Comment by taldcroft** [comment]
I haven't looked much in detail but this is a great idea to define the API with respect to Time missing values and scalars.  You may want to put `[skip ci]` in each commit until tests are passing locally.

**Comment by mhvk** [comment]
General discussion in #6509 - here just a note that if you want `np.isna[nt]` to work on `Time`, you'll need to define an `__array_ufunc__` to tell how it should treat `Time` instances. If you'd also want things like `np.nanmean` to work, one would likely also need to define `__array_function__` (which is planned a bit anyway, for `np.linspace` - see #8610.

All that said, it would be good if all these tests passed! (Except as @taldcroft noted, remove tests on private properties - those should remain implementation details - possibly including how to get `isnan` to work!)

**Comment by MaxNoe** [comment]
@mhvk yes, thanks for the pointers.

This is what I started with, but I have not yet a working version

**Comment by MaxNoe** [comment]
@mhvk I started working on this again. But defining `array_ufunc` is tricky. I defined it only for `isnan` and `isnat` but now stuff starts failing because what used to work, like:

`np.equal(time, 0)` now also calls ufunc.

Do I now have to implement every single numpy ufunc in `Time.__array_ufunc`. There has to be a better way.

**Comment by mhvk** [comment]
@MaxNoe - indeed, every ufunc has to be dealt with. There is a mixin to let all operators go via the same path, but I'm not sure that is too helpful in this case. 

That said, I actually hadn't appreciated any ufuncs worked at all on `Time`! Are there any tests that are actually failing?

But on reflection/inspection, I see what happens: the ufunc creates an array of Time scalars and let those subtract:
```
In [9]: t = Time(np.arange(50000., 50003.), format='mjd')                                                                                                                

In [10]: np.subtract(t, t)                                                                                                                                   

Out[10]: 
array([<TimeDelta object: scale='tai' format='jd' value=0.0>,
       <TimeDelta object: scale='tai' format='jd' value=0.0>,
       <TimeDelta object: scale='tai' format='jd' value=0.0>],
      dtype=object)
```

So, dealing with those will make things better...  (right now, for 10 elements `t==t` is 100 times faster than `np.equal(t, t)`).

In any case, my sense would be to first of all think of ufuncs where we're adding new functionality, and then worry about those that we've broken.

**Comment by pllim** [comment]
Hello. I converted your PR to a draft PR due to #10706 .

**Comment by pllim** [comment]
Sorry for the duplicate bot posts. I fixed the bug upstream, hopefully.

**Pull request closed without merging**